# Prüfungsleitung Maschnielles Lernen: Gruppe 8

Isolet download: https://datahub.io/machine-learning/isolet

### Aufgabe 1

#### 1) Data Loading and Splitting 

In [1]:
# Import isolet.csv File
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as sk

# import
df_raw_isolet = pd.read_csv("./raw_isolet.csv", delimiter=",")

# changing column name 'class' to 'letter' 'cause class is a reservated keyword in python 
df_raw_isolet.rename(columns={'class': 'letter'}, inplace=True)
# removing char ' from letter
df_raw_isolet['letter'] = df_raw_isolet['letter'].replace({'\'': ''}, regex=True)

##### Einfache Datenbetrachtung

In [2]:
print("Einträge in isolet.csv", len(df_raw_isolet))
print("Gruppierte Einträge:", df_raw_isolet.groupby(['letter']).size())

Einträge in isolet.csv 7797
Gruppierte Einträge: letter
1     300
10    300
11    300
12    300
13    299
14    300
15    300
16    300
17    300
18    300
19    300
2     300
20    300
21    300
22    300
23    300
24    300
25    300
26    300
3     300
4     300
5     300
6     298
7     300
8     300
9     300
dtype: int64


Bei der Betrachtung der Anzahl ist aufällig, dass in der Regel pro Buchstabe 300 Einträge aufgezeichnet wurden. Da aber len(df) nur 7797 Einträge beinhaltet, gibt es einige Ausnahmen. <br>
Ausnahmen: Buchstabe 13 mit 299 Einträgen und Buchstabe 6 mit 298 Einträgen <br><br>
Insgesamt bietet die sehr ausgewogene Verteilung der jeweiligen Buchstaben eine sehr gute Basis dies in dem Data Splitting zu betrachten. So wird für jeden Buchstaben ein Testsatz von je 75 Einträge bestimmt. Formel: ⎡Anzahl Einträge pro Buchstabe * 0,25⎤

##### Data Splitting

In [3]:
df_training = df_raw_isolet.copy()
df_test = pd.DataFrame(data=None, columns=df_raw_isolet.columns)
numberOfTestEntries = 75

for i in range(26):
    df_test = pd.concat([df_test, df_training[df_training.letter.str.match("^" + str(i+1) + "$")].head(numberOfTestEntries)])
df_training.drop(df_test.index, inplace=True)

# check
print("Number of rows in df_raw_isolet:", len(df_raw_isolet))
print("Number of rows in df_training:", len(df_training))
print("Number of rows in df_test", len(df_test))
if len(df_training) + len(df_test) == len(df_raw_isolet):
    print("OK: Check passed!")
else: 
    print("WARNING: The number of lines do not match! Diffrence:", len(df_training) + len(df_test) - len(df_raw_isolet))    


Number of rows in df_raw_isolet: 7797
Number of rows in df_training: 5847
Number of rows in df_test 1950
OK: Check passed!


In [4]:
# heatmap can not be pushed to git. calculate it on your local machine

# plt.figure(figsize=(150, 100))
# fig = sns.heatmap(df_training.corr(), annot=True, cmap=sns.cm.rocket_r)

#### 2) Merkmalsreduktion

#### 3) Implementierung der Klassifikatoren

#### 4) Evaluation